src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,180 @@
+
+/-- Adding a point in the affine span and a vector in the spanning
+subspace produces a point in the affine span. -/
+lemma vadd_mem_span_points_of_mem_span_points_of_mem_vector_span {s : set P} {p : P} {v : V}
+    (hp : p ∈ span_points k V s) (hv : v ∈ vector_span k V s) : v +ᵥ p ∈ span_points k V s :=
+begin
+  rcases hp with ⟨p2, ⟨hp2, ⟨v2, ⟨hv2, hv2p⟩⟩⟩⟩,
+  rw [hv2p, vadd_assoc],
+  use [p2, hp2, v + v2, (vector_span k V s).add hv hv2, rfl]
+end
+
+/-- Subtracting two points in the affine span produces a vector in the
+spanning subspace. -/
+lemma vsub_mem_vector_span_of_mem_span_points_of_mem_span_points {s : set P} {p1 p2 : P}
+    (hp1 : p1 ∈ span_points k V s) (hp2 : p2 ∈ span_points k V s) :
+  p1 -ᵥ p2 ∈ vector_span k V s :=
+begin
+  rcases hp1 with ⟨p1a, ⟨hp1a, ⟨v1, ⟨hv1, hv1p⟩⟩⟩⟩,
+  rcases hp2 with ⟨p2a, ⟨hp2a, ⟨v2, ⟨hv2, hv2p⟩⟩⟩⟩,
+  rw [hv1p, hv2p, vsub_vadd_eq_vsub_sub V (v1 +ᵥ p1a), vadd_vsub_assoc, add_comm, add_sub_assoc],
+  have hv1v2 : v1 - v2 ∈ (vector_span k V s).carrier,
+  { apply (vector_span k V s).add hv1,
+    rw ←neg_one_smul k v2,
+    exact (vector_span k V s).smul (-1 : k) hv2 },
+  refine (vector_span k V s).add _ hv1v2,
+  unfold vector_span,
+  change p1a -ᵥ p2a ∈ submodule.span k (vsub_set V s),
+  have hp1p2 : p1a -ᵥ p2a ∈ vsub_set V s, { use [p1a, hp1a, p2a, hp2a] },
+  have hp1p2s : vsub_set V s ⊆ submodule.span k (vsub_set V s) := submodule.subset_span,
+  apply set.mem_of_mem_of_subset hp1p2 hp1p2s
+end
+
+end affine_space
+
+open add_torsor affine_space
+
+section affine_subspace
+
+variables (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V] [vector_space k V]
+          [S : affine_space k V P]
+include S
+
+/-- An `affine_subspace k V P` is a subset of an `affine_space k V P`
+that has an affine space structure induced by a corresponding subspace
+of the `vector_space k V`. -/
+structure affine_subspace :=
+(carrier : set P)
+(direction : subspace k V)
+(nonempty : carrier.nonempty)
+(add : ∀ (p : P) (v : V), p ∈ carrier → v ∈ direction.carrier → v +ᵥ p ∈ carrier)
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction.carrier)
+
+/-- The whole affine space as a subspace of itself. -/
+def univ_affine_subspace : affine_subspace k V P :=
+{ carrier := set.univ,
+  direction := submodule.span k set.univ,
+  nonempty := set.nonempty_iff_univ_nonempty.1 S.nonempty,
+  add := λ p v hp hv, set.mem_univ _,
+  sub := begin
+    intros p1 p2 hp1 hp2,
+    apply set.mem_bInter,
+    intros x hx,
+    rw set.mem_set_of_eq at hx,
+    exact set.mem_of_mem_of_subset (set.mem_univ _) hx
+  end }
+
+instance : inhabited (affine_subspace k V P) := ⟨univ_affine_subspace k V P⟩
+
+/-- The affine span of a nonempty set of points is the smallest affine
+subspace containing those points. (Actually defined here in terms of
+spans in vector spaces.) -/
+def affine_span (s : set P) (h : s.nonempty) : affine_subspace k V P :=
+{ carrier := span_points k V s,
```

Comment:
If `span_points` is "deprecated" from now on, I think you should flag that in a comment at the definition of `span_points`.

Reply:
I don't think it's deprecated, but in most cases (if you know the set is nonempty), you probably want an `affine_subspace` and so probably want this function instead. I expanded the command on `span_points`.

Reply:
Ok, if `span_points` is not deprecated, then you probably want a lemma that says that the `coe` of `affine_span` to a set is `span_points`.
There should probably also be a `has_mem` instance for affine subspaces. Besides that, a lemma that says that elements of `s` are contained in `affine_span s`.

Reply:
Added that coercion, `has_mem` instance and lemmas.